File: 2019/LevelUp_1May20/lecture001_Recursion/l001_basic.cpp
#include <iostream>
#include <vector>
using namespace std;

void printIncr(int a, int b)
{
    if (a == b + 1)
        return;

    cout << a << " ";
    printIncr(a + 1, b);
}

void printDec(int a, int b)
{
    if (a == b + 1)
        return;

    printDec(a + 1, b);
    cout << a << " ";
}

void printIncreDec(int a, int b)
{
    if (a == b + 1)
        return;

    if (a % 2 == 0)
        cout << a << " hi " << endl;
    printIncreDec(a + 1, b);
    if (a % 2 != 0)
        cout << a << " bye " << endl;
}

//recursionWithArray.==========================================

void recursionWithArray()
{
    // int n;
    // cin >> n;
    // vector<int> arr(n, 0);
    // for (int i = 0; i < n; i++)
    //     cin >> arr[i];

    vector<int> arr{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
}

int maximum(vector<int> &arr, int idx)
{
    if (idx == arr.size() - 1)
        return arr[idx];

    return max(arr[idx], maximum(arr, idx + 1));
}

int Minimum(vector<int> &arr, int idx)
{
    if (idx == arr.size() - 1)
        return arr[idx];

    return min(arr[idx], Minimum(arr, idx + 1));
}

bool find(vector<int> &arr, int idx, int data)
{
    if (idx == arr.size())
        return false;
    if (arr[idx] == data)
        return true;
    return find(arr, idx + 1, data);
}

//questionSet.==================================================

vector<string> words = {
    ":;/",
    "abc",
    "def",
    "ghi",
    "jkl",
    "mno",
    "pqrs",
    "tuv",
    "wx",
    "yz",
    "&*%",
    "#@$",
};
vector<string> keyPad(string str)
{
    if (str.length() == 0)
    {
        vector<string> base;
        base.push_back("");
        return base;
    }

    char ch = str[0];
    string nstr = str.substr(1);
    string word = words[ch - '0'];

    vector<string> smallAns = keyPad(nstr);
    vector<string> myAns;

    for (string s : smallAns)
    {
        for (int i = 0; i < word.length(); i++)
            myAns.push_back(word[i] + s);
    }
    return myAns;
}

int keyPad(string ques, string ans)
{
    if (ques.length() == 0)
    {
        cout << ans << endl;
        return 1;
    }

    char ch = ques[0];
    string nques = ques.substr(1);
    string word = words[ch - '0'];

    int count = 0;
    for (int i = 0; i < word.length(); i++)
        count += keyPad(nques, ans + word[i]);

    return count;
}

int keyPad02(string ques, int idx, string ans)
{
    if (idx == ques.length())
    {
        cout << ans << endl;
        return 1;
    }

    char ch = ques[idx];
    string word = words[ch - '0'];

    int count = 0;
    for (int i = 0; i < word.length(); i++)
        count += keyPad02(ques, idx + 1, ans + word[i]);

    if (idx + 1 < ques.length())
    {
        int num = (ch - '0') * 10 + (ques[idx + 1] - '0');
        if (num == 10 || num == 11)
        {
            word = words[num];
            for (int i = 0; i < word.length(); i++)
                count += keyPad02(ques, idx + 2, ans + word[i]);
        }
    }
    return count;
}

//PathType.================================================

vector<string> mazepath_HVD(int sr, int sc, int er, int ec)
{
    if (sr == er && sc == ec)
    {
        vector<string> base;
        base.push_back("");
        return base;
    }

    vector<string> myAns;

    if (sr + 1 <= er)
    {
        vector<string> Verti = mazepath_HVD(sr + 1, sc, er, ec);
        for (string s : Verti)
            myAns.push_back("V" + s);
    }

    if (sc + 1 <= ec)
    {
        vector<string> Hori = mazepath_HVD(sr, sc + 1, er, ec);
        for (string s : Hori)
            myAns.push_back("H" + s);
    }

    if (sc + 1 <= ec && sr + 1 <= er)
    {
        vector<string> Diag = mazepath_HVD(sr + 1, sc + 1, er, ec);
        for (string s : Diag)
            myAns.push_back("D" + s);
    }
}

vector<string> mazepathMulti_HVD(int sr, int sc, int er, int ec)
{
    if (sr == er && sc == ec)
    {
        vector<string> base;
        base.push_back("");
        return base;
    }

    vector<string> myAns;

    for (int jump = 1; sr + jump <= er; jump++)
    {
        vector<string> Verti = mazepathMulti_HVD(sr + jump, sc, er, ec);
        for (string s : Verti)
            myAns.push_back("V" + to_string(jump) + s);
    }

    for (int jump = 1; sc + jump <= ec; jump++)
    {
        vector<string> Hori = mazepathMulti_HVD(sr, sc + jump, er, ec);
        for (string s : Hori)
            myAns.push_back("H" + to_string(jump) + s);
    }

    for (int jump = 1; sr + jump <= er && sc + jump <= ec; jump++)
    {
        vector<string> Diag = mazepathMulti_HVD(sr + jump, sc + jump, er, ec);
        for (string s : Diag)
            myAns.push_back("D" + to_string(jump) + s);
    }
}

int mazepath_HVD(int sr, int sc, int er, int ec, string ans)
{
    if (sr == er && ec == ec)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    if (sr + 1 <= er)
        count += mazepath_HVD(sr + 1, sc, er, ec, ans + "V");
    if (sc + 1 <= ec)
        count += mazepath_HVD(sr, sc + 1, er, ec, ans + "H");
    if (sr + 1 <= er && sc + 1 <= ec)
        count += mazepath_HVD(sr + 1, sc + 1, er, ec, ans + "D");

    return count;
}

int mazepath_MultiHVD(int sr, int sc, int er, int ec, string ans)
{
    if (sr == er && sc == ec)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    for (int jump = 1; sr + jump <= er; jump++)
        count += mazepath_MultiHVD(sr + jump, sc, er, ec, ans + "V" + to_string(jump));

    for (int jump = 1; sc + jump <= ec; jump++)
        count += mazepath_MultiHVD(sr, sc + jump, er, ec, ans + "H" + to_string(jump));

    for (int jump = 1; sr + jump <= er && sc + jump <= ec; jump++)
        count += mazepath_MultiHVD(sr + jump, sc + jump, er, ec, ans + "D" + to_string(jump));

    return count;
}

vector<vector<int>> dirA = {{0, -1}, {-1, -1}, {-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}};
vector<string> dirS = {"L", "N", "U", "E", "R", "S", "D", "W"};

int floodFill(int sr, int sc, int er, int ec, vector<vector<int>> &board, string ans) // 0 free cell and 1 is blocked cell
{
    if (sr == er && sc == ec)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0; //mark.
    board[sr][sc] = 1;
    for (int d = 0; d < dirA.size(); d++)
    {

        for (int rad = 1; rad <= board.size(); rad++)
        {
            int x = sr + rad * dirA[d][0];
            int y = sc + rad * dirA[d][1];
            if (x >= 0 && y >= 0 && x < board.size() && y < board[0].size() && board[x][y] == 0)
            {
                count += floodFill(x, y, er, ec, board, ans + dirS[d]);
            }
        }
    }
    board[sr][sc] = 0; //unmark
    return count;
}

void pathType()
{
    int n = 3, m = 5;
    // cout << mazepath_MultiHVD(0, 0, n - 1, m - 1, "") << endl;
    vector<vector<int>> board(3, vector<int>(3, 0));
    cout << floodFill(0, 0, 2, 2, board, "") << endl;
}

void questionSet()
{
    // vector<string> ans = keyPad("489");
    // cout << keyPad("489", "");
    cout << keyPad("1011489", "");

    // for (string s : ans)
    //     cout << s << " ";
}

void basics()
{
    int a, b;
    cin >> a >> b;

    // printIncr(a, b);
    // printDec(a, b);
    printIncreDec(a, b);
}

void solve()
{
    // basics();
    // questionSet();
    pathType();
}

int main()
{
    solve();
    return 0;
}

File: 2019/LevelUp_1May20/lecture001_Recursion/l001_basic.java
import java.util.Scanner;
import java.util.ArrayList;

public class l001_basic{
    public static Scanner scn=new Scanner(System.in);
    public static void main(String[] args){
      solve();
    }

    public static void solve(){
        // basic();
        questionSet();
    }

    public static void basic(){
        int a=scn.nextInt();
        int b=scn.nextInt();
        // System.out.println(fact(a));
        System.out.println(power(a,b));
    }

    public static int fact(int n){
        if(n<=1) return 1;
        return fact(n-1)*n;
    }

    public static int power(int a,int b){
        if(b==0) return 1;
        return power(a,b-1)*a;
    }

    public static int powerBtr(int a,int b){
        if(b==0) return 1;
        
        int halfAns=powerBtr(a,b/2);
        halfAns*=halfAns;
        
        return halfAns*(b%2!=0?a:1);
    }

    public static int calls(int n) {  // n=5
        if (n <= 1) {
            System.out.println("base: " + n);
            return n + 1;
        }

        int count = 0;

        System.out.println("Pre: " + n);
        count += calls(n - 1);

        System.out.println("In: " + n);

        count += calls(n - 2);
        System.out.println("Post: " + n);

        return count + 3;
    }

    //recursionWithArray.==========================================

    public static void recursionWithArray(){
        // int n=scn.nextInt();
        // int[] arr=new int[n];
        // for(int i=0;i<n;i++) arr[i]=scn.nextInt();

        int[] arr={1,2,3,4,5,6,7,8,9,10};
    }


    public static void display(int[] arr,int idx){
        if(idx==arr.length){
            return;
        }

        System.out.println(arr[idx]);
        display(arr,idx+1);
    } 

    public static int multiplyAllEle(int[] arr,int idx){
        if(idx==arr.length) return 1;
        return multiplyAllEle(arr,idx+1)*arr[idx];
    }

    public static int firstIndex(int[] arr,int idx,int data){
        if(idx==arr.length) return -1;

        if(arr[idx]==data) return idx;
        return firstIndex(arr,idx+1,data);
    }

    
    public static int lastIndex(int[] arr,int idx,int data){
        if(idx==arr.length) return -1;
        
        int ans=lastIndex(arr,idx+1,data);
        if(ans!=-1) return ans;

        return arr[idx] == data ? idx : -1; 

    }

    
    public static int[] AllIndex(int[] arr,int idx,int data,int count){
        if(idx==arr.length){
            return new int[count];
        }

        if(arr[idx]==data) count++;
        
        int[] ans=AllIndex(arr,idx+1,data,count);
        
        if(arr[idx]==data)
            ans[count-1]=idx;

        return ans;
    }

    //questionSet.==================================================

    public static void questionSet(){
        //    ArrayList<String> ans=permutation("abcd");
        //    for(String s: ans) System.out.println(s);
        // System.out.println(subseq("abc",""));
        System.out.println(permutation("aba",""));
        
    }

    public static ArrayList<String> subsequence(String str){
        if(str.length()==0){
            ArrayList<String> base=new ArrayList<>();
            base.add("");
            return base;
        }

        char ch=str.charAt(0);
        String nstr=str.substring(1); 
        
        ArrayList<String> smallAns=subsequence(nstr);
        ArrayList<String> myAns=new ArrayList<>();

        // for(int i=0;i<smallAns.size();i++){
        //     String s=smallAns.get(i);
        //     myAns.add(s);
        //     myAns.add(ch+s);            
        // }

        for(String s:smallAns){
            myAns.add(s);
            myAns.add(ch+s);
        }

        return myAns;
    }

    public static int subseq(String str,String ans){
        if(str.length()==0){
            System.out.println(ans);
            return 1;
        }

        char ch=str.charAt(0);
        String nstr=str.substring(1); 
        int count=0;

        count+=subseq(nstr,ans+ch);
        count+=subseq(nstr,ans);
        
        return count;
    }

    public static ArrayList<String> permutation(String str){
        if(str.length()==1){
            ArrayList<String> base=new ArrayList<>();
            base.add(str);
            return base;
        }

        char ch=str.charAt(0);
        String nstr=str.substring(1); 
        
        ArrayList<String> smallAns=permutation(nstr);
        ArrayList<String> myAns=new ArrayList<>();

        for(String s:smallAns){
            for(int i=0;i<=s.length();i++){
                String ans=s.substring(0,i) + ch + s.substring(i);
                myAns.add(ans);
            }
        }
        return myAns;
    }

    public static int permutation(String str,String ans){
        if(str.length()==0){
            System.out.println(ans);
            return 1;
        }

        int count=0;
        for(int i=0;i<str.length();i++){
            char ch=str.charAt(i);
            String nstr=str.substring(0,i) + str.substring(i+1);
            count+=permutation(nstr,ch + ans);
        }
        return count;
    }

    public static int permutationUnique(String str,String ans){
        if(str.length()==0){
            System.out.println(ans);
            return 1;
        }

        int count=0;
        boolean[] vis=new boolean[26];
        for(int i=0;i<str.length();i++){
            char ch=str.charAt(i);
            if(!vis[ch-'a']){
                vis[ch-'a']=true;
                String nstr=str.substring(0,i) + str.substring(i+1);
                count+=permutationUnique(nstr,ch + ans);
            }
        }
        return count;
    }
}

int[][] dirA = {{0, -1}, {-1, -1}, {-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}};
String[] dirS = {"L", "N", "U", "E", "R", "S", "D", "W"};

class pair{
    String path;
    int len;
    
    pair(String path_,int len_){
        path=path_;
        len=len_;
    }
}

pair shortestPath(int sr, int sc, int er, int ec,int[][] board){
    if(sr==er && sc==ec){
        return new pair("",0);
    }

    board[sr][sc]=1;
    pair myAns=new pair("",board.length * board[0].length);
    
    for(int d=0;d<dirA.length;d++){
        int x = sr+ dirA[d][0];
        int y = sc+ dirA[d][1];
        
        if(x >= 0 && y >= 0 && x < board.length && y < board[0].length && board[x][y]==0){
            pair smallAns=shortestPath(x,y,er,ec,board);
            if(smallAns + 1 < myAns.len ){
                myAns.len = smallAns.len + 1;
                myAns.path = dirS[d] + smallAns.path;
            }
        }
    }

    board[sr][sc]=0;
    return myAns;
}